File: src/main/java/skywolf46/bsl/server/BungeeSwitchListener.java
package skywolf46.bsl.server;

import net.md_5.bungee.BungeeCord;
import net.md_5.bungee.api.plugin.Plugin;
import net.md_5.bungee.config.Configuration;
import net.md_5.bungee.config.ConfigurationProvider;
import net.md_5.bungee.config.YamlConfiguration;
import skywolf46.bsl.global.BungeeSwitchListenerCore;
import skywolf46.bsl.global.BungeeVariables;
import skywolf46.bsl.global.abstraction.enums.Side;
import skywolf46.bsl.global.api.BSLCoreAPI;
import skywolf46.bsl.global.impl.packets.PacketPayload;
import skywolf46.bsl.global.util.ByteBufUtility;
import skywolf46.bsl.server.command.ConnectedInfoCommand;
import skywolf46.bsl.server.configuration.BungeeSwitchConfiguration;
import skywolf46.bsl.server.netty.hijack.BungeeConnectionHijacker;

import java.io.File;
import java.io.IOException;
import java.nio.channels.Channel;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.HashMap;

public class BungeeSwitchListener extends Plugin {
    private static BungeeSwitchConfiguration config;
    private static HashMap<Integer, Channel> serverChannels = new HashMap<>();


    @Override
    public void onEnable() {
        BungeeSwitchListenerCore.init(Side.SERVER);
        BSLCoreAPI.writer().printText("BSLCore initialized to server (Bungeecord)");
        BungeeCord.getInstance().getPluginManager().registerCommand(this, new ConnectedInfoCommand());
        File fx = new File(getDataFolder(), "config.yml");
        if (!fx.exists()) {
            BSLCoreAPI.writer().printText("Configuration file not exists. Generating new one.");
//            saveResource("config.yml", true);
            try {
                fx.getParentFile().mkdirs();
                fx.createNewFile();
                Files.copy(getResourceAsStream("bungee/config.yml"), fx.toPath(), StandardCopyOption.REPLACE_EXISTING);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        BSLCoreAPI.writer().printText("Parsing configuration...");
        Configuration configuration = null;
        try {
            configuration = ConfigurationProvider.getProvider(YamlConfiguration.class).load(new File(getDataFolder(), "config.yml"));
        } catch (IOException e) {
            BSLCoreAPI.writer().printError("Configuration parsing failed.");
        }
        config = new BungeeSwitchConfiguration(configuration);
        BungeeSwitchListenerCore.setIdentify(config.getAuthID());
        BungeeConnectionHijacker.hijack();
        BSLCoreAPI.getPacket(BungeeVariables.PACKET_GLOBAL_PAYLOAD)
                .attachListener((c, packet) -> {
                    PacketPayload payload = (PacketPayload) packet;
                    System.out.println(ByteBufUtility.readString(payload.getBuffer()));
                    System.out.println(((PacketPayload) packet).getBuffer().readInt());
                });
    }

    public static BungeeSwitchConfiguration getConfig() {
        return config;
    }


}

File: server/src/main/kotlin/skywolf46/bsl/server/BungeeSwitchListener.kt
package skywolf46.bsl.server

import net.md_5.bungee.api.plugin.Plugin
import net.md_5.bungee.config.Configuration
import net.md_5.bungee.config.ConfigurationProvider
import net.md_5.bungee.config.YamlConfiguration
import skywolf46.bsl.core.BSLCore
import skywolf46.bsl.core.impl.BSLServerHost
import skywolf46.bsl.core.security.permissions.SecurityPermissions
import skywolf46.bsl.server.impl.BungeeSyncProvider
import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption


class BungeeSwitchListener : Plugin() {
    companion object {
        var port = 44019
        val permissionMap = mutableMapOf<String, Pair<String, List<SecurityPermissions>>>()
        lateinit var inst: BungeeSwitchListener
            private set
    }

    override fun onEnable() {
        inst = this
        BSLCore.isServer = true
        BSLCore.init(getResourceAsStream("system.properties"))
        BSLCore.scanAll(file)
        BSLCore.changeSyncProvider(BungeeSyncProvider())
        with(File(dataFolder, "config.yml")) {
            if (!exists()) {
                parentFile.mkdirs()
                createNewFile()
                Files.copy(getResourceAsStream("bungee/config.yml"), toPath(), StandardCopyOption.REPLACE_EXISTING)
            }
            val configuration: Configuration =
                ConfigurationProvider.getProvider(YamlConfiguration::class.java).load(this)
            port = configuration.getInt("Server Port", port)
            if (configuration.getSection("System Codes") != null) {
                val sector = configuration.getSection("System Codes")
                for (x in sector.keys) {
                    val permSection = sector.getSection(x) ?: continue
                    if (!permSection.contains("Name") || !permSection.contains("Permission"))
                        continue
                    val permissions = mutableListOf<SecurityPermissions>()
                    for (perm in permSection.getStringList("Permission")) {
                        try {
                            permissions += SecurityPermissions.valueOf(perm)
                        } catch (_: Exception) {
                            // Ignore exception
                        }
                    }
                    permissionMap[x] = permSection.getString("Name") to permissions
                    println("BSL-Host | Registered permission ${permSection.getString("Name")} to $permissions")
                }
            }
            BSLServerHost(port)
            println("BSL-Host | Host server starting on $port")
        }
    }

    override fun onDisable() {
        BSLServerHost.host?.stopServer()
    }
}